File: utils/src/main/java/io/github/aemogie/timble/utils/logging/FileLoggerOutput.java
package io.github.aemogie.timble.utils.logging;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import org.jetbrains.annotations.Nullable;

import java.io.*;
import java.nio.file.Paths;

public class FileLoggerOutput extends LoggerOutput {
	private final String path;
	private final FileWriter file;
	
	public FileLoggerOutput(Logger.Level logLevel, String pattern, @Nullable JsonObject config) {
		super(logLevel, pattern, config);
		if (this.config == null) {
			Logger.SYS_ERR.printf("Could not locate config for \"%s\"! Please check your \"META-INF/timble-logger.json\"%n", getClass().getSimpleName());
			System.exit(-1);
		}
		JsonElement configPath = this.config.get("path");
		if (configPath == null) {
			Logger.SYS_ERR.printf("Could not locate config value \"path\" for \"%s\"! Please check your \"META-INF/timble-logger.json\"%n", getClass().getSimpleName());
			System.exit(-1);
		}
		path = configPath.getAsString();
		FileWriter tempFile = null;
		try {
			tempFile = new FileWriter(path);
		} catch (IOException e) {
			int i = path.lastIndexOf('/') + 1;
			File directory = new File(path.substring(0, i));
			if (!directory.exists() && directory.mkdir()) {
				try {
					tempFile = new FileWriter(path);
				} catch (IOException ioException) {
					Logger.SYS_ERR.println("Could not create log file at: " + Paths.get(path).toAbsolutePath());
				}
			}
		}
		file = tempFile;
	}
	
	@Override
	protected boolean print(String[] msg) {
		try {
			for (String s : msg) file.write(s);
			file.flush();
			return true;
		} catch (IOException e) {
			Logger.SYS_ERR.println("Unable to write to log file.");
			return false;
		}
	}
	
	@Override
	protected String[] colourise(String[] out, Logger.Level level) {
		return out;
	}
	
	@Override
	boolean destroy() {
		try {
			file.close();
			return true;
		} catch (IOException e) {
			Logger.SYS_ERR.println("Unable to close log file - " + Paths.get(path).toAbsolutePath());
			return false;
		}
	}
}

File: utils/src/main/kotlin/io/github/aemogie/timble/utils/logging/FileLoggerOutput.kt
package io.github.aemogie.timble.utils.logging

import io.github.aemogie.timble.utils.console.STD_ERR
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.io.BufferedWriter
import java.io.File
import java.io.FileWriter
import java.io.IOException
import kotlin.system.exitProcess

@Suppress("unused") //used via reflection, from the json config
class FileLoggerOutput(config: JsonObject) : LoggerOutput(config) {

	private val path: String = config["path"].let {
		if (it == null) {
			STD_ERR.write("Could not locate config value \"path\" for \"${this::class.simpleName}\"!\n".toByteArray())
			STD_ERR.write("Please check your \"${CONFIG_PATH}\"\n".toByteArray())
			exitProcess(-1)
		} else it
	}.jsonPrimitive.content

	private val file: BufferedWriter = BufferedWriter(FileWriter(File(path).apply {
		parentFile.apply { if (!exists()) mkdirs() }
	}))

	override fun print(text: String) = try {
		file.write(text)
	} catch (e: IOException) {
		STD_ERR.write("Unable to write to log file \"$path\"\n".toByteArray())
		STD_ERR.write("${e.message}\n".toByteArray())
		STD_ERR.flush()
	}

	override fun destroy() = try {
		file.close()
	} catch (e: IOException) {
		STD_ERR.write("Unable to close log file \"$path\"\n".toByteArray())
		STD_ERR.write("${e.message}\n".toByteArray())
		STD_ERR.flush()
	}

	override fun format(record: LogRecord, current: String): String = record.instant.toEpochMilli().let {
		"[%tH:%tM:%tS] [%5s|%5s] [%s] %s%n".format(
			it, it, it,
			record.thread.name,
			record.level,
			record.caller.className,
			current
		)
	}
}